packages/sanity/src/core/hooks/useListener.ts

```diff
@@ -0,0 +1,242 @@
+      const currentDocuments = new Map(state.documents)
+      currentDocuments.set(receivedDocument._id, receivedDocument)
+
+      return {
+        ...state,
+        documents: currentDocuments,
+      }
+    }
+
+    case 'DOCUMENT_DELETED': {
+      const currentDocuments = new Map(state.documents)
+      currentDocuments.delete(action.id)
+
+      return {
+        ...state,
+        documents: currentDocuments,
+      }
+    }
+
+    case 'DOCUMENT_UPDATED': {
+      const updateDocument = action.payload
+      const id = updateDocument._id as string
+      const currentDocuments = new Map(state.documents)
+      currentDocuments.set(id, updateDocument)
+
+      return {
+        ...state,
+        documents: currentDocuments,
+      }
+    }
+
+    default:
+      return state
+  }
+}
+
+interface ListenerOptions<T> {
+  /**
+   * Groq query to listen to
+   */
+  query: string
+  client: SanityClient | null
+}
+const INITIAL_STATE = {documents: new Map()}
+const LISTEN_OPTIONS: ListenOptions = {
+  events: ['welcome', 'mutation', 'reconnect'],
+  includeResult: true,
+  visibility: 'query',
+}
+
+export function useListener<T extends SanityDocument>({
+  query,
+  client,
+}: ListenerOptions<T>): {
+  documents: T[]
+  error: Error | null
+  loading: boolean
+  dispatch: Dispatch<documentsReducerAction<T>>
+} {
+  const [state, dispatch] = useReducer(documentsReducer, INITIAL_STATE)
+  const [loading, setLoading] = useState<boolean>(true)
+  const [error, setError] = useState<Error | null>(null)
+
+  const didInitialFetch = useRef<boolean>(false)
+
+  const initialFetch$ = useCallback(() => {
+    if (!client) {
+      return of(null) // emits null and completes if no client
+    }
+    return client.observable.fetch<T[]>(query).pipe(
+      timeout(10000), // 10s timeout
+      map((res) => {
+        dispatch({type: 'DOCUMENTS_SET', payload: res})
+        didInitialFetch.current = true
+        setLoading(false)
+      }),
+      retry({
+        count: 2,
+        delay: 1000,
+      }),
+      catchError((err) => {
+        if (err.name === 'TimeoutError') {
+          console.error('Fetch operation timed out:', err)
+        }
+        setError(err)
+        return of(null) // ensure stream completion even on error
+      }),
+    )
+  }, [client, query])
+
+  const handleListenerEvent = useCallback(
+    (event: ListenEvent<Record<string, T>>) => {
+      // Fetch all documents on initial connection
+      if (event.type === 'welcome' && !didInitialFetch.current) {
+        // Do nothing here, the initial fetch is done in the useEffect below
+        initialFetch$()
+      }
+
+      // The reconnect event means that we are trying to reconnect to the realtime listener.
+      // In this case we set loading to true to indicate that we're trying to
+      // reconnect. Once a connection has been established, the welcome event
+      // will be received and we'll fetch all bundles again (above)
```

Comment:
Perhaps remove references to bundles in here too

Reply:
Thanks for catching!